### PR TITLE
[rl] grpo chunked loss

### PR DESCRIPTION
**Pull request by pianpwk**
continuation of https://github.com/pytorch/torchtitan/pull/3710 originally authored by @DamianSzwichtenberg

### CODE REVIEW

File: torchtitan/experiments/rl/losses/grpo.py
Comment (felipemello1) [inline torchtitan/experiments/rl/losses/grpo.py:17]: not your fault, just making a note: GRPO inheriting from DAPO doesnt make sense, since GPRO is simpler than DAPO

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -338,6 +357,57 @@ def __init__(self, config: Config, *, compile_config: CompileConfig | None = Non
...
+    labels: torch.Tensor,
+) -> torch.Tensor:
+    """Compute per-position logprobs from logits and pre-shifted labels.
+
+    ``labels`` is pre-shifted per episode in the batcher
+    (``labels[i] = raw_token_ids[i+1]``), matching the pre-training
+    dataloader convention. No internal shift is needed.
+    Output shape matches input: ``[batch, seq_len]``.
+    """
+    if isinstance(logits, DTensor):
```
Comment (felipemello1) [inline torchtitan/components/loss.py:378]: noob question: `compute_logprobs` will be called inside of GPRO. Why do we need this check here but not in the GRPO loss?

File: torchtitan/components/loss.py [hunk 2]
```diff
@@ -500,46 +574,63 @@ class Config(BaseLoss.Config):
...
+        if (
+            hasattr(self.loss_fn, "global_vocab_size")
+            and getattr(self.loss_fn, "global_vocab_size") is None
+            and hasattr(lm_head, "out_features")
+        ):
+            self.loss_fn.global_vocab_size = lm_head.out_features
 
     def __call__(
         self,
         pred: torch.Tensor,
```
Comment (felipemello1) [inline torchtitan/components/loss.py:582]: not part of this PR, but i find "pred" misleading. This should be something like "hidden" or "lm_head_projection" or something like that. "pred" makes e think of logits/logprobs, not the raw head.

File: torchtitan/experiments/rl/losses/dapo.py [hunk 3]
```diff
@@ -29,44 +30,55 @@ class DAPOLoss(Configurable):
...
         advantages: torch.Tensor,
-        num_global_valid_tokens: int,
+        loss_mask: torch.Tensor,
     ) -> tuple[torch.Tensor, dict[str, torch.Tensor]]:
         """Compute the per-token clip-higher surrogate loss.
 
         Args:
-            trainer_logprobs: [B, L] log π_θ(a_t | s_t) from the current policy.
-            generator_logprobs: [B, L] log π_old(a_t | s_t) from the sampling policy.
+            pred: [B, L, V] logits from the current policy.
```
Comment (felipemello1) [inline torchtitan/experiments/rl/losses/dapo.py]: this is wrong. Preds is the hidden layer output. I think that we should rename so mistakes like this dont happen again. "hidden" is a good alternative.

File: torchtitan/components/loss.py [hunk 4]
```diff
@@ -575,6 +666,7 @@ def _chunk(t):
                 c.detach().requires_grad_(requires_grad) for c in _chunk(hidden_states)
             ]
             label_chunks = list(_chunk(labels))
+            input_chunks = {key: _chunk(value) for key, value in loss_inputs.items()}
```
Comment (felipemello1) [inline torchtitan/components/loss.py]: i think that we should apply chunk only if its a tensor, else pass as it is.

File: torchtitan/components/loss.py [hunk 5]
```diff
@@ -338,6 +357,57 @@ def __init__(self, config: Config, *, compile_config: CompileConfig | None = Non
...
+
+def reduce_metric_value(
+    key: str,
+    current: torch.Tensor | None,
+    value: torch.Tensor,
+) -> torch.Tensor:
+    if current is None:
+        return value
+    if key.endswith(("/mean", "/frac", "_mean", "_frac")):
+        return current + value
```
Comment (felipemello1) [inline torchtitan/components/loss.py]: we should add a docstring that this expects the metrics to be already normalized by global valid tokens, and thats why summing it makes sense for mean computation.

File: torchtitan/components/loss.py [hunk 6]
```diff
@@ -614,16 +707,14 @@ def _chunk(t):
...
+                chunk_loss, chunk_metrics = self.loss_fn(
+                    logits, label_chunk, global_valid_tokens, **chunk_inputs
                 )
-                if global_valid_tokens is not None:
-                    chunk_loss = chunk_loss / global_valid_tokens
-                if (
-                    get_spmd_backend() == "spmd_types"
-                ):  # V -> P reinterpret, after exiting local region
+                for key, value in chunk_metrics.items():
+                    metrics[key] = reduce_metric_value(key, metrics.get(key), value)
```
Comment (felipemello1) [inline torchtitan/components/loss.py]: maybe change the signature to:

```metrics = reduce_metric_value(chunk_metrics)```

But is "reduce" the right word here? We still need to do an all_gather later on the `loss_mesh`. Maybe combine and add a small comment/docstring that the all_gather is still necessary?

File: torchtitan/components/loss.py (hunk 1 above)
Comment (pianpwk) [reply torchtitan/components/loss.py:378]: hm I don't think I understood this one , this ensures the logits are allgathered if vocab-sharded, in case we were using TP.

Maybe this placement is ok because it's the only usage of pred in DAPO

File: torchtitan/components/loss.py (hunk 6 above)
Comment (pianpwk) [reply torchtitan/components/loss.py]: comments should be addressed, thanks for the patience with review :)

File: torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py [hunk 7]
```diff
@@ -766,3 +767,13 @@ def rl_grpo_qwen3_0_6b_varlen_batch_invariant() -> RLTrainer.Config:
...
         ),
     )
+
+
+def rl_grpo_qwen3_0_6b_varlen_batch_invariant_chunked() -> RLTrainer.Config:
+    """On-policy GRPO for Qwen3-0.6B with chunked loss."""
+    config = rl_grpo_qwen3_0_6b_varlen_batch_invariant()
+    config.trainer = dataclasses.replace(
+        config.trainer,
+        loss=ChunkedLossWrapper.Config(num_chunks=8, loss_fn=GRPOLoss.Config()),
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py]: instead of introducing a new one, we should change every config to use chunked loss

File: torchtitan/components/loss.py [hunk 8]
```diff
@@ -338,6 +357,70 @@ def __init__(self, config: Config, *, compile_config: CompileConfig | None = Non
...
+        # Gather vocab-sharded TP logits before computing per-token logprobs.
+        placements = tuple(
+            Replicate()
+            if isinstance(p, Shard) and p.dim in (-1, logits.ndim - 1)
+            else p
+            for p in logits.placements
+        )
+        logits = logits.redistribute(placements=placements).to_local()
+
+    B, S, V = logits.shape
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: maybe `B, L, V` to be consistent with e.g. MoE

File: torchtitan/experiments/rl/losses/dapo.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/experiments/rl/losses/dapo.py]: @felipemello1 what is wrong?
I think it should be `logits`.

File: torchtitan/components/loss.py [hunk 9]
```diff
@@ -294,11 +313,11 @@ def __call__(
         pred: torch.Tensor,
         labels: torch.Tensor,
         global_valid_tokens: float | None = None,
-    ) -> torch.Tensor:
+    ) -> tuple[torch.Tensor, dict[str, torch.Tensor]]:
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:323]: add docstring for returns

File: torchtitan/components/loss.py [hunk 10]
```diff
@@ -338,6 +357,70 @@ def __init__(self, config: Config, *, compile_config: CompileConfig | None = Non
...
 
+def compute_logprobs(
+    logits: torch.Tensor,
+    labels: torch.Tensor,
+) -> torch.Tensor:
+    """Compute per-position logprobs from logits and pre-shifted labels.
+
+    ``labels`` is pre-shifted per episode in the batcher
+    (``labels[i] = raw_token_ids[i+1]``), matching the pre-training
+    dataloader convention. No internal shift is needed.
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: not necessary to mention pre-shifting

File: torchtitan/components/loss.py [hunk 11]
```diff
@@ -338,6 +357,70 @@ def __init__(self, config: Config, *, compile_config: CompileConfig | None = Non
...
+    B, S, V = logits.shape
+    return -F.cross_entropy(
+        logits.float().reshape(B * S, V),
+        labels.reshape(B * S),
+        reduction="none",
+        ignore_index=IGNORE_INDEX,
+    ).reshape(B, S)
+
+
+def combine_chunk_metrics(
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]:
```suggestion
def _combine_chunk_metrics(
```

Comment (tianyu-l) [reply torchtitan/components/loss.py]: also I feel this should belong to `ChunkedLossWrapper`

File: torchtitan/models/common/decoder.py [hunk 12]
```diff
@@ -155,10 +154,12 @@ def update_from_config(
                 )
 
             loss_config = getattr(config, "loss", None)
-            if isinstance(loss_config, (ChunkedCELoss.Config, CrossEntropyLoss.Config)):
+            if hasattr(loss_config, "global_vocab_size"):
                 # TODO(pianpwk): Move this into config_registry entries. This
```
Comment (tianyu-l) [inline torchtitan/models/common/decoder.py]: This is exactly why we should move global_vocab_size to config construction time.
- you are passing global_vocab_size blindly to ChunkedLossWrapper, despite the inner loss_fn is LossParalleCrossEntropy or not, but only that inner loss_fn requires the global_vocab_size.

File: torchtitan/components/loss.py [hunk 13]
```diff
@@ -500,46 +587,68 @@ class Config(BaseLoss.Config):
         global_vocab_size: int | None = None
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: shouldn't exist here

File: torchtitan/components/loss.py [hunk 14]
```diff
@@ -500,46 +587,68 @@ class Config(BaseLoss.Config):
...
     ):
-        self.fn: LossFunction = cross_entropy_loss
-        self._maybe_compile(compile_config)
         self.num_chunks = config.num_chunks
+        if config.global_vocab_size is not None and hasattr(
+            config.loss_fn, "global_vocab_size"
+        ):
+            config.loss_fn.global_vocab_size = (
+                config.global_vocab_size
+            )  # pyrefly: ignore[missing-attribute]
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: this, looks a bit awkward

File: torchtitan/components/loss.py [hunk 15]
```diff
@@ -500,46 +587,68 @@ class Config(BaseLoss.Config):
...
         """Set the lm_head module. Must be called before the first __call__."""
         self.lm_head = lm_head
+        if (
+            hasattr(self.loss_fn, "global_vocab_size")
+            and getattr(self.loss_fn, "global_vocab_size") is None  # noqa: B009
+            and hasattr(lm_head, "out_features")
+        ):
+            self.loss_fn.global_vocab_size = (
+                lm_head.out_features
+            )  # pyrefly: ignore[missing-attribute]
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: similar

File: torchtitan/components/loss.py [hunk 16]
```diff
@@ -500,46 +587,68 @@ class Config(BaseLoss.Config):
...
         num_chunks = self.num_chunks
         lm_head = self.lm_head
-        assert lm_head is not None, "Set lm_head before calling ChunkedCELoss"
+        assert lm_head is not None, "Set lm_head before calling ChunkedLossWrapper"
+        if self.global_vocab_size is not None and hasattr(
+            self.loss_fn, "global_vocab_size"
+        ):
+            self.loss_fn.global_vocab_size = (
+                self.global_vocab_size
+            )  # pyrefly: ignore[missing-attribute]
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: again

**Review by tianyu-l — COMMENTED** [review]
We can continue to work here until approval, but would you be OK if we do final merge in https://github.com/pytorch/torchtitan/pull/3710 so that @DamianSzwichtenberg gets credit, despite we took over for the rush?

File: torchtitan/experiments/rl/losses/dapo.py (hunk 3 above)
Comment (felipemello1) [reply torchtitan/experiments/rl/losses/dapo.py]: sorry, you right. It was late and this threw me off

image

But i just realized that `lm_head` is optional, so we cannot call it `hidden` either.

Correction:
Return it to "pred" in the ChunkedLossWrapper
And change it to "logits" in DAPO/GRPO

do you agree @tianyu-l ?

**Comment by pianpwk** [comment]
> We can continue to work here until approval, but would you be OK if we do final merge in #3710 so that @DamianSzwichtenberg gets credit, despite we took over for the rush?

ah of course, sorry I was too dumb to figure out pushing to that branch

File: torchtitan/models/common/decoder.py (hunk 12 above)
Comment (pianpwk) [reply torchtitan/models/common/decoder.py]: uh switched this , but i'm not sure if the mechanism of extracting model specs is what you had in mind

**Review by tianyu-l — APPROVED** [review]
looks great
